feat(dashboards): Adds pagination and table sort support to issue widget viewer

edwardgou-sentry opened the pull request:
Extends the Widget Viewer pagination and column sort functionality to also support Issue Widgets.

narsaynorath commented on static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx:
```diff
@@ -37,7 +39,32 @@ type Props = {
+          return sortField
+            ? {
+                ...location,
+                query: {...location.query, [WidgetViewerQueryField.SORT]: sortField},
```
Should we be negating the sort here if we are sorting by this field? Currently I'm unable to reverse the sorting.

edwardgou-sentry replied:
Hey Nar, Issues API doesn't support sorting in ascending order at this moment 😞 .
The api only supports a limited set of options, namely descending Events, descending Users, First Seen, and Last Seen (none of these in reverse order).
Due to this, we are only able to allow limited sorting for Issue Widgets in the Viewer.

narsaynorath commented on static/app/components/modals/widgetViewerModal/widgetViewerTableCell.tsx:
```diff
@@ -37,7 +39,32 @@ type Props = {
+        align={align}
+        title={<StyledTooltip title={column.name}>{column.name}</StyledTooltip>}
+        direction={widget.queries[0].orderby === sortField ? 'desc' : undefined}
+        canSort
```
Can we use `sortField` here to set `canSort={!!sortField}` so we don't have to do the ternary in `generateSortLink`?

From what I understand if you can't sort, then it won't render with a clickable header

edwardgou-sentry replied:
Yup, updated canSort and generateSortLink!

priscilawebdev commented on static/app/components/modals/widgetViewerModal.tsx:
```diff
@@ -150,27 +154,49 @@ function WidgetViewerModal(props: Props) {
+                    <StyledPagination
+                      pageLinks={pageLinks}
+                      onCursor={(nextCursor, _path, _query, delta) => {
+                        let nextPage: number | undefined = isNaN(pagination.page)
```
can `nextPage` get also an `undefined` value?

edwardgou-sentry replied:
Shouldn't be possible to get an `undefined` value here, I've updated the type to reflect this!

priscilawebdev commented on tests/js/spec/components/modals/widgetViewerModal.spec.tsx:
```diff
@@ -432,5 +471,60 @@ describe('Modals -> WidgetViewerModal', function () {
+      // Need to manually set the new router location and rerender to simulate the sortable column click
+      initialData.router.location.query = {modalSort: ['freq']};
+      rerender(
+        <div style={{padding: space(4)}}>
```
do we need this div element?

edwardgou-sentry replied:
Not needed! Updated!

narsaynorath reviewed (approved):
Looks good to me

One question I had about the sorting though, does it make sense to highlight the LifetimeEvents/LifetimeUsers headers if it's only doing freq and user? e.g. here's a screenshot where it's showing it's descending on LifetimeEvents but it's not actually doing that
![Screen Shot 2022-03-08 at 9 42 35 AM](https://user-images.githubusercontent.com/22846452/157261782-eb60ca88-e711-4aae-896f-cf5dcd437183.png)

edwardgou-sentry commented:
> Looks good to me
> 
> One question I had about the sorting though, does it make sense to highlight the LifetimeEvents/LifetimeUsers headers if it's only doing freq and user? e.g. here's a screenshot where it's showing it's descending on LifetimeEvents but it's not actually doing that

Good question, I added the freq and user sort fields to LifetimeEvents and LifetimeUsers since they were somewhat related, but I think it might cause some confusion for users since it's not directly sorting the Lifetime columns, as you pointed out. I've updated to remove it for now and will consult with Jasmin to see what we want to do.

priscilawebdev reviewed (approved):
lgtm
